### PR TITLE
QR factorization

### PR DESCRIPTION
**Pull request by nullhook**
## Proposed changes

Add QR factorization primitive

TODO:
- [x] impl for other dtypes
- [ ] tests
- [ ] python bindings

potential tests:
1. reconstruction
2. orthogonality
3. upper triangular 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [ ] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/backend/metal/primitives.cpp
Comment (awni) [inline mlx/backend/metal/primitives.cpp:631]: I think it's better to throw here with a "Metal QR factorization not yet implemented" rather than implicitly go to the CPU implementation. That way people know what they are getting.

File: mlx/linalg.cpp
Comment (awni) [inline mlx/linalg.cpp:9]: nit: mlx/primitives

**Review by awni — COMMENTED** [review]
This looks awesome so far! I left two minor comments, let us know when it's ready for full review.

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [inline mlx/backend/common/qrf.cpp:117]: @awni what's the status on multi array output PR?

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [inline mlx/backend/common/qrf.cpp:24]: not sure if we need an implementation for fp64 here because mlx doesn't support it?
for dtype complex64 we could use CGEQRF and CORMQR.

cc @awni

File: mlx/backend/common/qrf.cpp
Comment (awni) [reply mlx/backend/common/qrf.cpp:117]: Its probably at least a week out realistically. It's a big change so we'll have to test it thoroughly before we can merge it to main.

Comment (awni) [reply mlx/backend/common/qrf.cpp:117]: Are you blocked by that / what's your plan for working around it?

File: mlx/backend/common/qrf.cpp
Comment (awni) [reply mlx/backend/common/qrf.cpp:24]: We don't need to support fp64 yet. If there is complex64 that would be nice to add. For other floating point types (fp16 and bf16), I would just through for now.

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [reply mlx/backend/common/qrf.cpp:113]: it would be ideal to have multi-array output when merging this PR. however, we can work towards getting it close to ready, and once your PR merges, i can rebase.

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [reply mlx/backend/common/qrf.cpp:25]: makes sense

File: tests/linalg_qr_tests.cpp
Comment (nullhook) [inline tests/linalg_qr_tests.cpp:29]: you can see there's a slight mismatch in R matrix. the issue is mainly because lapack accepts col-major layout and mlx arrays are row-major. if I try transpose A before calling `triu`, i get the right output. however, i know transposing isn't ideal as mentioned in another thread.

i don't see any other way than manually re-arranging the memory layout to col-major and near the output change it back to row-major. ideas? @awni

**Comment by awni** [comment]
@nullhook you might want to rebase this as we have landed multi-output primitives. Are you still interested in contributing the QR factorization?

**Comment by nullhook** [comment]
@awni yes, i will rebase it soon.

**Comment by nullhook** [comment]
rebased

File: tests/linalg_qr_tests.cpp
Comment (nullhook) [reply tests/linalg_qr_tests.cpp:32]: cc @awni ideas?

File: mlx/backend/common/qrf.cpp
Comment (awni) [inline mlx/backend/common/qrf.cpp:69]: Why are you making scalar arrays here? Why not just use a `float`?

File: mlx/backend/common/qrf.cpp
Comment (awni) [reply mlx/backend/common/qrf.cpp]: Does the routine take a transpose flag? If so you could trick it and give it the row major matrix and say it's transposed (which is the same as a column major matrix).

Comment (nullhook) [reply mlx/backend/common/qrf.cpp]: are you referring to the lapack routines? both routines, xgeqrf and xorgqr does not take a transpose flag.

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [reply mlx/backend/common/qrf.cpp:57]: is there a reason to go through the allocator? rather than constructing an array on stack like above.

Comment (awni) [reply mlx/backend/common/qrf.cpp:57]: 1. An array is not a stack object, it is a very dynamic heap-based object which involves multiple allocations
2. But the main problem is you should never construct compute graphs inside a primitive's evaluation. What you are doing is building a new graph while the outer graph is running, so it is just a bug.

File: mlx/backend/common/qrf.cpp
Comment (awni) [inline mlx/backend/common/qrf.cpp:26]: Also I'm curious, why is this "xgeqrf" instead of `sgeqrf` which is the reference in the docs for single precision:

https://www.netlib.org/lapack/single/sgeqrf.f

Comment (nullhook) [reply mlx/backend/common/qrf.cpp:26]: internally it is calling sgeqrf. 

there are variants of 'sgeqrf', such as 'dgeqrf' for double precision and 'cgeqrf' for complex numbers. the above template, the 'x' in 'xgeqrf' serves as a placeholder that can be defined for a specific type T. 

basically, the same template can be used for different data types by simply substituting the appropriate type-specific LAPACK routine

File: mlx/backend/common/qrf.cpp
Comment (nullhook) [reply mlx/backend/common/qrf.cpp:57]: ah, that make sense. i guess you create intermediate buffers through the allocator like this then?

`Buffer b = allocator::malloc_or_wait(...);`

Comment (awni) [reply mlx/backend/common/qrf.cpp:57]: Precisely! 👍

**Comment by nullhook** [comment]
i kept it simple to only one dtype implementation, which is float. perhaps other dtypes can be implemented in a separate PR, if needed by the community?

@awni this is ready for review.

**Comment by awni** [comment]
I think this is in pretty good shape now.

- Compiles on linux
- Tests clear
- Multiple matrices work
- Python bindings + tests

Still no support for non-square matrices but maybe that is a follow up..

**Review by awni — APPROVED** [review]
Good to go for now. We can add non-square matrices as a follow up. Thanks for your work on this one!!

File: mlx/linalg.cpp
Comment (adhulipa) [inline mlx/linalg.cpp:192]: @nullhook @awni out of curiosity: what's the scope of work to support non-square?  It seems like the lapack funcs `sgeqrf` and `sorgqr` used here already support it. 

IIUC, for a non-square matrix A with M rows and N columns process is:
1. Call `sgeqrf` to compute the QR factorization A = Q*R, where Q is now an MxM  matrix, and R is an MxN upper trapezoidal matrix.
2. Call`sorgqr` to generate the MxM matrix Q from the reflectors (similar to current impl)

So I supposed the scope is to define some helper logic to calculate min(M, N) for estimating size of the trap matrix R; and consequently memory allocations for the various arrays.  (And perhaps a few other related details around the for-loops halting conditions etc)

Im curious if that's the right line of thinking? OR do you see anything major I could be missing?

Comment (awni) [reply mlx/linalg.cpp:192]: That sounds about right to me

**Pull request merged**
